dipy/denoise/denspeed_backup.pyx

```diff
@@ -0,0 +1,261 @@
+        int mm = arr.shape[3]              # n_DWIs
+        int c_patch_radius = patch_radius
+        int nn = (2 * c_patch_radius + 1) ** 3   # number of voxels in local kernel
+        int half_nn = int(nn/2.)
+        int rr = mm if (mm<nn) else nn     # dimension of the square PCA matrix
+
+
+        # for memory usage in OPENMP loops. We create a temp memory for each slice for each OPENMP thread.
+        # There should be a better way to do this.
+
+        # data matrix
+        double[:, :, ::1] X = np.zeros([arr.shape[2], nn, mm], dtype=np.double)
+        # eigenvector matrix
+        double[:, :, ::1] C = np.zeros([arr.shape[2], rr, rr], dtype=np.double)
+        # diagonal eigenvalue matrix
+        double[:, :, ::1] diag_eigmat = np.zeros([arr.shape[2], rr, rr],
+                                                 dtype=np.double)
+        # eigenvalue array
+        double[:, ::1] W = np.zeros([arr.shape[2], rr], dtype=np.double)
+        # eigenvalue cummulative
+        double[::1] cum_W = np.zeros([rr], dtype = np.double)
+        double[::1] eigenV = np.zeros([rr], dtype = np.double)
+
+        # temporary variables used in LAPACK calls
+        double[:, ::1] temp0 = np.zeros([arr.shape[2], rr], dtype=np.double)
+        double[:, ::1] temp1 = np.zeros([arr.shape[2], rr], dtype=np.double)
+        double[:, ::1] temp2 = np.zeros([arr.shape[2], mm], dtype=np.double)
+        double[:, ::1] temp3 = np.zeros([arr.shape[2], mm], dtype=np.double)
+        int LWORK=2*(1 + 6 * rr + 2 * rr * rr)
+        double[:, ::1] WORK = np.zeros([arr.shape[2], LWORK], dtype=np.double)
+        int LIWORK=2*(3+5*rr)
+        int[:, ::1] IWORK = np.zeros([arr.shape[2], LIWORK], dtype=np.int32)
+
+        # variables for noise cutoff eigenvalue computations
+        double gamma, sigma_hat, sigma2, rhs, r0
+        int p_hat
+
+        # for memoryviewing the original images
+        double[:,:,:,:] arr_view = arr
+
+    threads_to_use = num_threads or all_cores
+    patch_size = 2 * patch_radius + 1
+    if patch_size **3 < mm:
+        return arr
+    print("cy patch_size", patch_size, nn)
+
+    if have_openmp:
+        openmp.omp_set_dynamic(0)
+        openmp.omp_set_num_threads(threads_to_use)
+
+    sizes[0] = arr.shape[0]
+    sizes[1] = arr.shape[1]
+    sizes[2] = arr.shape[2]
+
+    noise_arr = np.zeros([arr.shape[0],arr.shape[1],arr.shape[2]],dtype=np.float64)
+    denoised_arr = np.zeros(arr.shape, dtype=np.float64)
+
+    if mask is None:
+        # If mask is not specified, use the whole volume
+        mask = np.ones_like(arr, dtype=int)[..., 0]
+
+
+    cdef double [:, :, :, :] denoised_arr_view = denoised_arr
+    cdef double [:, :, :] noise_arr_view = noise_arr
+    cdef int [:, :, :] mask_arr_view = mask.astype(np.int32)
+
+    cdef int e = 0
+    cdef double tamp = 0.
+    cdef int m2 = 0
+
+    # OPENMP loop over slices
+    for k in prange(0, sizes[2], nogil=True, schedule=static):
```

Comment:
but isn't a loop over slices going to reduce the number of samples? Treating each slice separately, the voxels must be drawn from a slice (i.e. 2D), thus reducing the number of samples = voxels. Either it is required to go 'further out' to get voxels from further away in order to satisfy number_of_samples > number_of_features (in which case, the noise level is more likely to change, and the redundancy in the feature dimension is likely to be worse), of number_of_samples < number_of_features (which isn't ideally what we want). Therefore, I would not be surprised that this performs worse than the original version

Reply:
Sorry but that doesn't generally hold. If you use a 3x3x3 spatial neighborhood, this is 27 voxels, so if N is above 27 then clearly you would have more features than samples... This might explain the slightly worse performance? Would you not tune the spatial neighborhood to guarantee sample > features?

I think I understand your point about the par over slices now though.